File: app/src/main/java/zukkey/whiskey/com/arcoresamplefortechbookfest/AugmentedImageActivity.java
package zukkey.whiskey.com.arcoresamplefortechbookfest;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.view.Gravity;
import android.widget.Button;
import android.widget.Toast;
import com.google.ar.core.Anchor;
import com.google.ar.core.AugmentedImage;
import com.google.ar.core.AugmentedImageDatabase;
import com.google.ar.core.Config;
import com.google.ar.core.Frame;
import com.google.ar.core.Session;
import com.google.ar.core.TrackingState;
import com.google.ar.core.exceptions.UnavailableApkTooOldException;
import com.google.ar.core.exceptions.UnavailableArcoreNotInstalledException;
import com.google.ar.core.exceptions.UnavailableSdkTooOldException;
import com.google.ar.sceneform.AnchorNode;
import com.google.ar.sceneform.ArSceneView;
import com.google.ar.sceneform.math.Quaternion;
import com.google.ar.sceneform.math.Vector3;
import com.google.ar.sceneform.rendering.ViewRenderable;
import com.google.ar.sceneform.ux.ArFragment;
import com.google.ar.sceneform.ux.TransformableNode;
import java.io.IOException;
import java.io.InputStream;
import java.util.Collection;

public class AugmentedImageActivity extends AppCompatActivity {
  private ViewRenderable textViewRenderable;
  private ArFragment arFragment;
  private Session session;
  private Boolean sessionConfigured = false;
  private ArSceneView arSceneView;
  private Boolean isAttachedModel = false;

  public static Intent createIntent(Context context) {
    return new Intent(context, AugmentedImageActivity.class);
  }

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_augumented_image);
    arFragment = (ArFragment) getSupportFragmentManager().findFragmentById(R.id.ar_fragment);

    if (arFragment != null) {
      // デフォルトだとPlaneの検出が始まるが、Planeを検出しない
      arFragment.getPlaneDiscoveryController().hide();
      arFragment.getPlaneDiscoveryController().setInstructionView(null);
      arSceneView = arFragment.getArSceneView();

      arFragment.getArSceneView().getScene().addOnUpdateListener(frameTime -> {
        Frame frame = arFragment.getArSceneView().getArFrame();
        Collection<AugmentedImage> updatedAugmentedImages = frame.getUpdatedTrackables(AugmentedImage.class);

        for (AugmentedImage img : updatedAugmentedImages) {
          if (img.getTrackingState() == TrackingState.TRACKING) {
            if (img.getName().contains("eure") && !isAttachedModel) {
              setUp3DModel(img.createAnchor(img.getCenterPose()));
            }
          }
        }
      });
    }

  }

  @Override
  protected void onResume() {
    super.onResume();
    if (session == null) {
      try {
        session = new Session(this);
      } catch (UnavailableArcoreNotInstalledException | UnavailableApkTooOldException | UnavailableSdkTooOldException e) {
        e.printStackTrace();
      }
      sessionConfigured = true;
    }

    if (sessionConfigured) {
      configureSession();
      sessionConfigured = false;
    }
  }

  @Override
  protected void onPause() {
    super.onPause();
  }

  private void setUp3DModel(Anchor anchor) {
    ViewRenderable.builder()
        .setView(this, R.layout.item_text)
        .build()
        .thenAccept(renderable -> textViewRenderable = renderable)
        .exceptionally(
            throwable -> {
              Toast toast =
                  Toast.makeText(this, "Unable to load andy renderable", Toast.LENGTH_LONG);
              toast.setGravity(Gravity.CENTER, 0, 0);
              toast.show();
              return null;
            });

    if (textViewRenderable == null) {
      return;
    }

    AnchorNode anchorNode = new AnchorNode(anchor);
    anchorNode.setParent(arFragment.getArSceneView().getScene());

    TransformableNode text = new TransformableNode(arFragment.getTransformationSystem());
    text.setParent(anchorNode);
    text.setRenderable(textViewRenderable);
    text.getRotationController();
    text.getScaleController();
    text.select();
    text.setLocalRotation(Quaternion.lookRotation(Vector3.down() ,Vector3.up()));
    Button btn = (Button) textViewRenderable.getView();
    btn.setOnClickListener(view -> {
      Toast.makeText(this, "ボタンをタップしました！詳細へ遷移します", Toast.LENGTH_SHORT).show();
      startActivity(WebActivity.createIntent(this));
      finish();
    });
    isAttachedModel = true;
  }

  private void configureSession() {
    Config config = new Config(session);
    InputStream inputStream;
    try {
      inputStream = getAssets().open("sample_database.imgdb");
      AugmentedImageDatabase imageDatabase = AugmentedImageDatabase.deserialize(session, inputStream);
      config.setAugmentedImageDatabase(imageDatabase);
      config.setUpdateMode(Config.UpdateMode.LATEST_CAMERA_IMAGE);
      session.configure(config);
    } catch (IOException e) {
      Toast.makeText(this, "Don't configure Session", Toast.LENGTH_SHORT).show();
      e.printStackTrace();
    }
    arSceneView.setupSession(session);
  }
}

File: app/src/main/java/zukkey/whiskey/com/arcoresamplefortechbookfest/kotlin/AugmentedImageActivity.kt
package zukkey.whiskey.com.arcoresamplefortechbookfest.kotlin

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.Gravity
import android.widget.Button
import android.widget.Toast
import com.google.ar.core.*
import com.google.ar.core.exceptions.UnavailableApkTooOldException
import com.google.ar.core.exceptions.UnavailableArcoreNotInstalledException
import com.google.ar.core.exceptions.UnavailableSdkTooOldException
import com.google.ar.sceneform.AnchorNode
import com.google.ar.sceneform.ArSceneView
import com.google.ar.sceneform.math.Quaternion
import com.google.ar.sceneform.math.Vector3
import com.google.ar.sceneform.rendering.ViewRenderable
import com.google.ar.sceneform.ux.ArFragment
import com.google.ar.sceneform.ux.TransformableNode
import zukkey.whiskey.com.arcoresamplefortechbookfest.R
import zukkey.whiskey.com.arcoresamplefortechbookfest.java.AugmentedImageActivity
import zukkey.whiskey.com.arcoresamplefortechbookfest.java.WebActivity
import java.io.IOException
import java.io.InputStream


class AugmentedImageActivity: AppCompatActivity() {
  private var textViewRenderable: ViewRenderable? = null
  private var arFragment: ArFragment? = null
  private var session: Session? = null
  private var sessionConfigured: Boolean = false
  private var arSceneView: ArSceneView? = null
  private var isAttachedModel: Boolean = false

  fun createIntent(context: Context): Intent {
    return Intent(context, AugmentedImageActivity::class.java)
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_augumented_image)
    arFragment = supportFragmentManager.findFragmentById(R.id.ar_fragment) as ArFragment?

    if (arFragment != null) {
      // デフォルトだとPlaneの検出が始まるが、Planeを検出しない
      arFragment!!.planeDiscoveryController.hide()
      arFragment!!.planeDiscoveryController.setInstructionView(null)
      arSceneView = arFragment!!.arSceneView

      arFragment!!.arSceneView.scene.addOnUpdateListener { frameTime ->
        val frame = arFragment!!.arSceneView.arFrame
        val updatedAugmentedImages = frame.getUpdatedTrackables(AugmentedImage::class.java)

        for (img in updatedAugmentedImages) {
          if (img.trackingState == TrackingState.TRACKING) {
            if (img.name.contains("eure") && !isAttachedModel) {
              setUp3DModel(img.createAnchor(img.centerPose))
            }
          }
        }
      }
    }

  }

  override fun onResume() {
    super.onResume()
    if (session == null) {
      try {
        session = Session(this)
      } catch (e: UnavailableArcoreNotInstalledException) {
        e.printStackTrace()
      } catch (e: UnavailableApkTooOldException) {
        e.printStackTrace()
      } catch (e: UnavailableSdkTooOldException) {
        e.printStackTrace()
      }

      sessionConfigured = true
    }

    if (sessionConfigured) {
      configureSession()
      sessionConfigured = false
    }
  }

  private fun setUp3DModel(anchor: Anchor) {
    ViewRenderable.builder()
        .setView(this, R.layout.item_text)
        .build()
        .thenAccept { renderable -> textViewRenderable = renderable }
        .exceptionally { throwable ->
          val toast = Toast.makeText(this, "Unable to load andy renderable", Toast.LENGTH_LONG)
          toast.setGravity(Gravity.CENTER, 0, 0)
          toast.show()
          null
        }

    if (textViewRenderable == null) {
      return
    }

    val anchorNode = AnchorNode(anchor)
    anchorNode.setParent(arFragment!!.arSceneView.scene)

    val text = TransformableNode(arFragment!!.transformationSystem)
    text.setParent(anchorNode)
    text.renderable = textViewRenderable
    text.rotationController
    text.scaleController
    text.select()
    text.localRotation = Quaternion.lookRotation(Vector3.down(), Vector3.up())
    val btn = textViewRenderable!!.view as Button
    btn.setOnClickListener { view ->
      Toast.makeText(this, "ボタンをタップしました！詳細へ遷移します", Toast.LENGTH_SHORT).show()
      startActivity(WebActivity.createIntent(this))
      finish()
    }
    isAttachedModel = true
  }

  private fun configureSession() {
    val config = Config(session!!)
    val inputStream: InputStream
    try {
      inputStream = assets.open("sample_database.imgdb")
      val imageDatabase = AugmentedImageDatabase.deserialize(session!!, inputStream)
      config.augmentedImageDatabase = imageDatabase
      config.updateMode = Config.UpdateMode.LATEST_CAMERA_IMAGE
      session!!.configure(config)
    } catch (e: IOException) {
      Toast.makeText(this, "Don't configure Session", Toast.LENGTH_SHORT).show()
      e.printStackTrace()
    }

    arSceneView!!.setupSession(session)
  }
}
